sdk/src/program_error.rs

```diff
@@ -0,0 +1,129 @@
+use crate::instruction::InstructionError;
+use num_traits::ToPrimitive;
+
+/// Reasons the program may fail
+pub enum ProgramError {
+    /// CustomError allows programs to implement program-specific error types and see
+    /// them returned by the Solana runtime. A CustomError may be any type that is represented
+    /// as or serialized to a u32 integer.
+    ///
+    /// NOTE: u64 requires special serialization to avoid the loss of precision in JS clients and
+    /// so is not used for now.
+    CustomError(u32),
+    /// The arguments provided to a program instruction where invalid
+    InvalidArgument,
+    /// An instruction's data contents was invalid
+    InvalidInstructionData,
+    /// An account's data contents was invalid
+    InvalidAccountData,
+    /// An account's data was too small
+    AccountDataTooSmall,
+    /// An account's balance was too small to complete the instruction
+    InsufficientFunds,
+    /// The account did not have the expected program id
+    IncorrectProgramId,
+    /// A signature was required but not found
+    MissingRequiredSignature,
+    /// An initialize instruction was sent to an account that has already been initialized.
+    AccountAlreadyInitialized,
+    /// An attempt to operate on an account that hasn't been initialized.
+    UninitializedAccount,
+    /// The instruction expected additional account keys
+    NotEnoughAccountKeys,
+    /// Failed to borrow a reference to an account, already borrowed
+    AccountBorrowFailed,
+}
+
+/// 32bit representations of builtin program errors returned by the entry point
+const BUILTIN_ERROR_START: u32 = 0x8000_0000; // 31st bit set
+const INVALID_ARGUMENT: u32 = BUILTIN_ERROR_START + 1;
+const INVALID_INSTRUCTION_DATA: u32 = BUILTIN_ERROR_START + 2;
+const INVALID_ACCOUNT_DATA: u32 = BUILTIN_ERROR_START + 3;
+const ACCOUNT_DATA_TOO_SMALL: u32 = BUILTIN_ERROR_START + 4;
+const INSUFFICIENT_FUNDS: u32 = BUILTIN_ERROR_START + 5;
+const INCORRECT_PROGRAM_ID: u32 = BUILTIN_ERROR_START + 6;
+const MISSING_REQUIRED_SIGNATURES: u32 = BUILTIN_ERROR_START + 7;
+const ACCOUNT_ALREADY_INITIALIZED: u32 = BUILTIN_ERROR_START + 8;
+const UNINITIALIZED_ACCOUNT: u32 = BUILTIN_ERROR_START + 9;
+const NOT_ENOUGH_ACCOUNT_KEYS: u32 = BUILTIN_ERROR_START + 10;
+const ACCOUNT_BORROW_FAILED: u32 = BUILTIN_ERROR_START + 11;
```

Comment:
It'd be nice to use c-like enums if we need to preserve this primitive mapping. Maybe built-in errors can get pulled out into their own enum and then we would have:

```rust
InstructionError {
   Builtin(BuiltinError),
   Custom(u32),
}
```

Reply:
Wouldn't this become InstructionError::ProgramError(program_error::ProgramError) with examples like:

`InstructionError::Program(ProgramError::InvalidArgument)`
`InstructionError::Program(ProgramError::Custom(u32))`

Reply:
Yeah, I think that's better. Made that comment before you suggested the `InstructionError::Program` approach